### DISCUSSION TITLE
Bugs in AutoAttack implementation

### DISCUSSION THREAD

**Discussion by SignedQiu** [Trusted-AI/adversarial-robustness-toolbox#2109] [Q&A] [answered]
**Describe the bug**
There is a data type error in the auto pgd implementaion of autoattack.

**To Reproduce**
Simply run the AutoAttack using the default parameters in Mnist dataset and PyTorchClassifier (where i met the bug), the errors will occurs with "RuntimeError: Expected object of scalar type Long but got scalar type Byte for argument #2 'target' in call to _thnn_nll_loss_forward"

**Expected behavior**
I attempted to fix this bug, but i think the code  "grad = self.estimator.loss_gradient(x_k, y_batch) * (1 - 2 * int(self.targeted))" in the 455th line of the file auto_projected_gradient_descent.py is the key to solving this problem.

**Screenshots**
![use](https://user-images.githubusercontent.com/66827919/232747907-6177d289-7a4c-4b15-a456-916dc3a0f800.jpg)

![image](https://user-images.githubusercontent.com/66827919/232747743-77a50495-0b72-494e-9765-ebf90b1be14d.png)

**System information (please complete the following information):**
 - OS = Linux
 - Python version = 3.7.9
 - ART version or commit number  = 1.14.0
 - PyTorch = 1.8.2+cu111

**Comment by SignedQiu** [comment]
In the process of attempting to fix the vulnerability, I initially thought it was a simple data type error, and I completed the data type conversion at the corresponding position. However, I discovered a new bug. The input of the "target" parameter in the function "torch._C._nn.nll_loss" does not conform to the dimensional specification. The function "self.reduce_labels(y_preprocessed)" clearly checks the dimensions, but it appears that the check was bypassed due to some conditional settings. Therefore, I believe this is a system-level bug in the auto pgd implementation. I don't have any more time to fix it, so I am temporarily reporting the bug here.

**Comment by beat-buesser** [comment]
Hi @SignedQiu Thank you for using ART! What is the dtype of variable `y_batch`?

**Reply by SignedQiu** [reply]
Thank you for your response. Although I haven't figured out what caused the issue, I did find out that there were some problems with the cross_entropy function implemented by the code author. After changing the code in "auto_projected_gradient_descent.py" from this:
![image](https://user-images.githubusercontent.com/66827919/232807291-b5ad2ad8-f86a-42da-88fc-9f25cf6b578f.png)
to this:
![image](https://user-images.githubusercontent.com/66827919/232807574-aa30d766-6996-454c-8a6e-a254dafe3ea3.png)
everything works fine now!

**Reply by SignedQiu** [reply]
The dtype of y_batch is np.uint8. I have modified the code as above and didn't change the dtype of y_batch, everthing works fine now.

**Reply by beat-buesser** [reply]
What is the shape of your model's output layer?

**Reply by SignedQiu** [reply]
(batch_size, 10)

**Reply by beat-buesser** [reply]
Thank you. I have noticed that your example does not provide true labels in line `adv_inputs = attack.generate(x=x_test)`. In that case the attack uses your model's output as true label to attack. What is the dtype of the output of your model`?

**Reply by SignedQiu** [reply]
On GPU, the dtype of the output of the model is torch.float32.

**Reply by beat-buesser** [reply, marked as answer]
Hi @SignedQiu I think you have found a bug! Actually two. It looks like the order of prediction and label arguments is wrong in the custom loss class and the PyTorchClassifier does not recognize the custom loss class to set internal attributes correctly. I have opened issue #2116 and we'll include the solution in ART 1.14.1. Thank you for report this issue!

**Comment by beat-buesser** [comment]
I'm converting this issue into a discussion on the Discussion tab to continue.
